cartographer_grpc/framework/execution_context.h

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2017 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CARTOGRAPHER_GRPC_FRAMEWORK_EXECUTION_CONTEXT_H
+#define CARTOGRAPHER_GRPC_FRAMEWORK_EXECUTION_CONTEXT_H
+
+#include "cartographer/common/make_unique.h"
+#include "cartographer/common/mutex.h"
+#include "glog/logging.h"
+
+namespace cartographer_grpc {
+namespace framework {
+
+// Implementations of this class allow RPC handlers to share state among one
+// another. Using Server::SetExecutionContext(...) a server-wide
+// 'ExecutionContext' can be specified. This 'ExecutionContext' can be retrieved
+// by all implementations of 'RpcHandler' by calling
+// 'RpcHandler::GetContext<MyContext>()'.
+class ExecutionContext {
+ public:
+  // This non-movable, non-copyable class is used to broker access from various
+  // RPC handlers to the shared 'ExecutionContext'. Handles automatically lock
+  // the context they point to.
+  template <typename T>
+  class Handle {
+   public:
+    T* operator->() { return static_cast<T*>(execution_context_); }
+    Handle(cartographer::common::Mutex* lock,
+           ExecutionContext* execution_context)
+        : locker_(cartographer::common::make_unique<
+                  cartographer::common::MutexLocker>(lock)),
+          execution_context_(execution_context) {}
+    Handle(const Handle&) = delete;
+    Handle(Handle&&) = delete;
+
+   private:
+    std::unique_ptr<cartographer::common::MutexLocker> locker_;
+    ExecutionContext* execution_context_;
+  };
+  cartographer::common::Mutex* lock() { return &lock_; }
```

Comment:
Since the only usage is to get a Handle for an ExecutionContext, how about a method
`Handle<T> ExecutionContext::GetSynchronizedAccess()` (or a different name)? Is this possible?
This would have no more public API than needed.

Reply:
I can add that, but it would not be used at the moment. In RpcHandler I have to construct Synchronized via implicit construction using a list initializer, because Synchronized is not copyable. I would like to keep Synchronized non-copyable so people can't store it an that way sit on the lock.